src/git-repository-async.js

```diff
@@ -0,0 +1,873 @@
+  // changed. When a file is updated, reloaded, etc, and the status changes, this
+  // will be fired.
+  //
+  // * `callback` {Function}
+  //   * `event` {Object}
+  //     * `path` {String} the old parameters the decoration used to have
+  //     * `pathStatus` {Number} representing the status. This value can be passed to
+  //       {::isStatusModified} or {::isStatusNew} to get more information.
+  //
+  // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
+  onDidChangeStatus (callback) {
+    return this.emitter.on('did-change-status', callback)
+  }
+
+  // Public: Invoke the given callback when a multiple files' statuses have
+  // changed. For example, on window focus, the status of all the paths in the
+  // repo is checked. If any of them have changed, this will be fired. Call
+  // {::getPathStatus(path)} to get the status for your path of choice.
+  //
+  // * `callback` {Function}
+  //
+  // Returns a {Disposable} on which `.dispose()` can be called to unsubscribe.
+  onDidChangeStatuses (callback) {
+    return this.emitter.on('did-change-statuses', callback)
+  }
+
+  // Repository details
+  // ==================
+
+  // Public: A {String} indicating the type of version control system used by
+  // this repository.
+  //
+  // Returns `"git"`.
+  getType () {
+    return 'git'
+  }
+
+  // Public: Returns a {Promise} which resolves to the {String} path of the
+  // repository.
+  getPath () {
+    return this.repoPromise.then(repo => repo.path().replace(/\/$/, ''))
+  }
+
+  // Public: Returns a {Promise} which resolves to the {String} working
+  // directory path of the repository.
+  getWorkingDirectory () {
+    return this.repoPromise.then(repo => repo.workdir())
+  }
+
+  // Public: Returns a {Promise} that resolves to true if at the root, false if
+  // in a subfolder of the repository.
+  isProjectAtRoot () {
+    if (!this.project) return Promise.resolve(false)
+
+    if (!this.projectAtRoot) {
+      this.projectAtRoot = this.repoPromise
+        .then(repo => this.project.relativize(repo.workdir()))
+        .then(relativePath => relativePath === '')
+    }
+
+    return this.projectAtRoot
+  }
+
+  // Public: Makes a path relative to the repository's working directory.
+  relativize (_path, workingDirectory) {
+    // Cargo-culted from git-utils. The original implementation also handles
+    // this.openedWorkingDirectory, which is set by git-utils when the
+    // repository is opened. Those branches of the if tree aren't included here
+    // yet, but if we determine we still need that here it should be simple to
+    // port.
+    //
+    // The original implementation also handled null workingDirectory as it
+    // pulled it from a sync function that could return null. We require it
+    // to be passed here.
+    if (!_path || !workingDirectory) {
+      return _path
+    }
+
+    if (process.platform === 'win32') {
+      _path = _path.replace(/\\/g, '/')
```

Comment:
That's a good question. As the comment indicates, this was cargo-culted from `git-utils` so I don't have any context on it. @thedaniel any idea?